### PR TITLE
[Merged by Bors] - feat: For any `b`, there exists a set `s` of independent atoms such that `Sup s` is the complement of `b`

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/8475

* [`order.compactly_generated`@`210657c4ea4a4a7b234392f70a3a2a83346dfa90`..`e8cf0cfec5fcab9baf46dc17d30c5e22048468be`](https://leanprover-community.github.io/mathlib-port-status/file/order/compactly_generated?range=210657c4ea4a4a7b234392f70a3a2a83346dfa90..e8cf0cfec5fcab9baf46dc17d30c5e22048468be)
* [`order.directed`@`485b24ed47b1b7978d38a1e445158c6224c3f42c`..`e8cf0cfec5fcab9baf46dc17d30c5e22048468be`](https://leanprover-community.github.io/mathlib-port-status/file/order/directed?range=485b24ed47b1b7978d38a1e445158c6224c3f42c..e8cf0cfec5fcab9baf46dc17d30c5e22048468be)

---

- [x] depends on: #2543

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2543
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2543~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Order/CompactlyGenerated.lean [hunk 1]
```diff
@@ -520,70 +552,82 @@ instance (priority := 100) isAtomistic_of_complementedLattice [ComplementedLatti
...
+      ⟨⋃₀ c,
+        ⟨CompleteLattice.independent_unionₛ_of_directed hc2.directedOn fun s hs => (hc1 hs).1, ?_,
+          fun a ⟨s, sc, as⟩ => (hc1 sc).2.2 a as⟩,
+        fun _ => Set.subset_unionₛ_of_mem⟩
+  -- swap
+  -- · rw [supₛ_unionₛ, ← supₛ_image, DirectedOn.disjoint_supₛ_right]
+  --   · rintro _ ⟨s, hs, rfl⟩
+  --     exact (hc1 hs).2.1
+  --   · rw [directedOn_image]
+  --     exact hc2.directed_on.mono fun s t => supₛ_le_supₛ
```
Comment (YaelDillies) [inline Mathlib/Order/CompactlyGenerated.lean:578]: Weird error here. It complains that it cannot synthesize the placeholder?? The `swap, rw ...` is what is supposed to prove the placeholder, FYI.

Comment (kmill) [reply Mathlib/Order/CompactlyGenerated.lean:578]: Last I checked, `obtain` and `have` can't have placeholders in their bodies. You should be able to do
```lean
obtain ⟨s, ⟨s_ind, b_inf_Sup_s, s_atoms⟩, s_max⟩ := by
  refine zorn_subset ... etc etc ...
  rw ... etc etc ...
```
which is cleaner I think.

Comment (YaelDillies) [reply Mathlib/Order/CompactlyGenerated.lean:578]: But there's an example of it working on l. 443 of this same file!

Comment (kmill) [reply Mathlib/Order/CompactlyGenerated.lean:578]: I see. It seems to be some peculiarity with `obtain`. It works if you split it into a `have` then an `obtain`:
```
  have := zorn_subset
      { s : Set α | CompleteLattice.SetIndependent s ∧ Disjoint b (supₛ s) ∧ ∀ a ∈ s, IsAtom a }
      fun c hc1 hc2 =>
      ⟨⋃₀ c,
        ⟨CompleteLattice.independent_unionₛ_of_directed hc2.directedOn fun s hs => (hc1 hs).1, ?_,
          fun a ⟨s, sc, as⟩ => (hc1 sc).2.2 a as⟩,
        fun _ => Set.subset_unionₛ_of_mem⟩
  obtain ⟨s, ⟨s_ind, b_inf_Sup_s, s_atoms⟩, s_max⟩ := this
  swap
  · rw [supₛ_unionₛ, ← supₛ_image, DirectedOn.disjoint_supₛ_right]
    · rintro _ ⟨s, hs, rfl⟩
      exact (hc1 hs).2.1
    · rw [directedOn_image]
      exact hc2.directedOn.mono @fun s t => supₛ_le_supₛ
```

Comment (YaelDillies) [reply Mathlib/Order/CompactlyGenerated.lean]: Left a porting note.

File: Mathlib/Order/Directed.lean [hunk 2]
```diff
@@ -63,10 +63,17 @@ theorem directedOn_iff_directed {s} : @DirectedOn α r s ↔ Directed r (Subtype
 alias directedOn_iff_directed ↔ DirectedOn.directed_val _
 #align directed_on.directed_coe DirectedOn.directed_val
 
-theorem directedOn_range {f : β → α} : Directed r f ↔ DirectedOn r (Set.range f) := by
+theorem directedOn_range {f : ι → α} : Directed r f ↔ DirectedOn r (Set.range f) := by
   simp_rw [Directed, DirectedOn, Set.forall_range_iff, Set.exists_range_iff]
 #align directed_on_range directedOn_range
 
+-- porting note: This lemma was misplaced in mathlib
```
Comment (eric-wieser) [inline Mathlib/Order/Directed.lean:70]: What do you mean by this?

Comment (YaelDillies) [reply Mathlib/Order/Directed.lean:70]: The `alias` was declared in `order.compactly_generated`

Comment (eric-wieser) [reply Mathlib/Order/Directed.lean:70]:
```suggestion
-- porting note: This alias was misplaced in `order/compactly_generated.lean` in mathlib3
```

**Review by Vierkantor — APPROVED** [review]
Thanks!

bors merge

**Pull request closed without merging**